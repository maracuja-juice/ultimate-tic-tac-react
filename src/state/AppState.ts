import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
