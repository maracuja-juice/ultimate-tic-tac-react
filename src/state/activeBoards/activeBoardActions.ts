import { Point } from '../../util/Point';

export const SET_ACTIVE_BOARDS = 'activeBoardReducer/SetActiveBoard';
export const CALCULATE_ACTIVE_BOARDS = 'activeBoardReducer/CalculateActiveBoard';

export const setActiveBoards = ( boards: Point[], ) => ({
    type: SET_ACTIVE_BOARDS,
    payload: boards
});
export const calculateActiveBoards = ( lastMove: Point ) => ({
    type: CALCULATE_ACTIVE_BOARDS,
    payload: lastMove
});