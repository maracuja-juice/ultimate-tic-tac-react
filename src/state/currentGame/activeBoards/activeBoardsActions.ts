import { Point } from '../../../util/Point';

export const SET_ALLOWED_BOARDS = 'activeBoardsReducer/set-allowed-boards';
export const CALCULATE_ALLOWED_BOARDS = 'activeBoardsReducer/calculate-allowed-boards';

export const setAllowedBoards = ( boards: Point[], ) => ({
    type: SET_ALLOWED_BOARDS,
    payload: boards
});
export const calculateActiveBoards = ( lastMove: Point ) => ({
    type: CALCULATE_ALLOWED_BOARDS,
    payload: lastMove
});