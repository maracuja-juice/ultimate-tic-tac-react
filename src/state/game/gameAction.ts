import { Player } from '../AppState';
import { Point } from '../../util/Point';

export const CHANGE_PLAYER = 'game-reducer/player-changed';
export const PLAYER_MOVED = 'game/player-moved';
export const PLAYER_MOVED_VALID = 'game/player-moved-valid';
export const GAME_FINISHED = 'game/game-finished';
export const CHECK_GAME_FINISHED = 'game/check-game-finished';
export const RESTART_GAME = 'restart-game';

export const changePlayer = () => ({
    type: CHANGE_PLAYER
});

export const restartGame = () => ({
    type: RESTART_GAME
});

export const gameFinished = ( winningPlayer?: Player | null ) => ({
    type: GAME_FINISHED,
    payload: winningPlayer
});

export const checkGameFinished = (  ) => ({
    type: CHECK_GAME_FINISHED
});

export const playerMoved = ( boardPosition: Point, tilePosition: Point ): PlayerMovedAction => ({
    type: PLAYER_MOVED,
    payload: {
        boardPosition: boardPosition,
        tilePosition: tilePosition
    }
});

export interface PlayerMovedAction {
    type: string;
    payload: {
        boardPosition: Point
        tilePosition: Point
    };
}
