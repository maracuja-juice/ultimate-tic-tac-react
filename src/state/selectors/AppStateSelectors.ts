import { AnalysisGame, AppState, Player } from '../AppState';

export const getBoards = (state: AppState) => state.currentGame.board;

export const getCurrentPlayer = (state: AppState) =>
  state.currentGame.game.currentPlayer;

export const getWinningPlayer = (state: AppState) =>
  state.currentGame.game.winningPlayer;

export const getMoves = (state: AppState) => state.currentGame.moves;

export const getFinishedGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
      board: finishedGame.gameState,
      id: finishedGame.id!,
    };
  }

  return undefined;
};
