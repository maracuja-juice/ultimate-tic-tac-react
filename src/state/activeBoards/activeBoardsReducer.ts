import { Point } from '../../util/Point';
import { GenericAction } from '../AppState';
import { SET_ALLOWED_BOARDS } from './activeBoardsActions';
import { RESTART_GAME } from '../commonAction';

const getAllPoints = () => {
    let points: Point[] = [];

    for (let x = 0; x < 3; x++) {
        for (let y = 0; y < 3; y++) {
            points.push( {x, y} );
        }
    }

    return points;
};

const initialState: Point[] = getAllPoints();

const activeBoardsReducer = ( state = initialState, action: GenericAction ) => {
    switch (action.type) {
        case SET_ALLOWED_BOARDS: {
            return action.payload;
        }
        case RESTART_GAME: {
            return initialState;
        }
        default: {
            return state;
        }
    }
};

export default activeBoardsReducer;