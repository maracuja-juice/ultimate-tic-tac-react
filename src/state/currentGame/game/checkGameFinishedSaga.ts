import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction } from '../../AppState';
import { CHECK_GAME_FINISHED, gameFinished } from './gameAction';
import { getMoves } from '../../selectors/AppStateSelectors';
import { setAllowedBoards } from '../activeBoards/activeBoardsActions';
import { saveGameData } from '../../finishedGames/saveFinishedGameDataActions';
import { getFinishedGameData } from '../../selectors/FinishedGameStateSelectors';
import { TicTacToeGame } from '../../../util';

function* checkIfGameFinished(action: GenericAction) {
  const moves = yield select(getMoves);

  const winResult = new TicTacToeGame(moves).getWinResult();
  const isGameFinished = winResult.isFinished;
  const winningPlayer = winResult.winningPlayer;

  if (isGameFinished) {
    yield put(gameFinished(winningPlayer));
    yield put(setAllowedBoards([]));
    yield put(saveGameData(yield select(getFinishedGameData)));
  }
}

function* checkGameFinishedSaga() {
  yield takeEvery(CHECK_GAME_FINISHED, checkIfGameFinished);
}

export default checkGameFinishedSaga;
