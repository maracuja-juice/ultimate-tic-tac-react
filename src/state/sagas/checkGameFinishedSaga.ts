import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction, SmallBoardInformation } from '../AppState';
import { CHECK_GAME_FINISHED, gameFinished } from '../game/gameAction';
import { getBoards } from '../selectors/AppStateSelectors';
import { getWinResult } from '../../util/CheckBoard';

function* checkIfGameFinished( action: GenericAction ) {
    const boards = yield select( getBoards );

    // TODO: make this map more beautiful or change the params that the getWinResult function takes
    const mappedBoards = boards.map( (smallBoard: SmallBoardInformation) => {
        return {smallBoardPoint: smallBoard.point, value: smallBoard.value};
    });
    const winResult = getWinResult(mappedBoards);

    const isGameFinished = winResult.isFinished;
    const winningPlayer = winResult.winningPlayer;

    if (isGameFinished) {
        yield put( gameFinished( winningPlayer! ) );
    }
}

function* checkGameFinishedSaga() {
    yield takeEvery( CHECK_GAME_FINISHED, checkIfGameFinished );
}

export default checkGameFinishedSaga;