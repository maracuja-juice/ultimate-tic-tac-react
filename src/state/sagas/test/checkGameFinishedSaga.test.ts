import { expectSaga } from 'redux-saga-test-plan';
import checkGameFinishedSaga from '../checkGameFinishedSaga';
import { getBoards } from '../../selectors/AppStateSelectors';
import { select } from 'redux-saga/effects';
import { CHECK_GAME_FINISHED, GAME_FINISHED } from '../../game/gameAction';
import { Player } from '../../AppState';
import { SET_ALLOWED_BOARDS } from '../../activeBoards/activeBoardsActions';
import { circleFinishedBoardMock, crossFinishedBoardMock } from './finishedBoardMock';
import unfinishedBoardMock from './unfinishedBoardMock';

describe( 'checkGameFinishedSaga', () => {
    it( 'should dispatch gameFinished action with Circle and' +
        ' setAllowedBoards to no board when the game is finished',
        () => {
            return expectSaga( checkGameFinishedSaga )
                .provide( [
                              [select( getBoards ), circleFinishedBoardMock]
                          ] )
                .put( {type: GAME_FINISHED, payload: Player.Circle} )
                .put( {type: SET_ALLOWED_BOARDS, payload: []} )
                .dispatch( {type: CHECK_GAME_FINISHED} )
                .run();
        } );

    it( 'should dispatch gameFinished action with Cross and setAllowedBoards to no board ' +
        'if the game is finished',
        () => {
        return expectSaga( checkGameFinishedSaga )
            .provide( [
                          [select( getBoards ), crossFinishedBoardMock]
                      ] )
            .put( {type: GAME_FINISHED, payload: Player.Cross} )
            .put( {type: SET_ALLOWED_BOARDS, payload: []} )
            .dispatch( {type: CHECK_GAME_FINISHED} )
            .run();
    } );

    it( 'should dispatch no action if the game is not finished', () => {
        return expectSaga( checkGameFinishedSaga )
            .provide( [
                          [select( getBoards ), unfinishedBoardMock]
                      ] )
            .dispatch( {type: CHECK_GAME_FINISHED} )
            .run();
    } );

    // if more put effects happen: this catches it + this checks for the order
    it( 'should match snapshot', () => {
        return expectSaga( checkGameFinishedSaga )
            .provide( [
                          [select( getBoards ), circleFinishedBoardMock]
                      ] )
            .dispatch( {type: CHECK_GAME_FINISHED} )
            .run()
            .then( ( result ) => {
                expect( result.toJSON() ).toMatchSnapshot();
            } );
    } );

} );