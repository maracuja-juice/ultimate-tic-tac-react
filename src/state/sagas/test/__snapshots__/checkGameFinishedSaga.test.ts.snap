// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkGameFinishedSaga should match snapshot 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "game/check-game-finished",
          "@@redux-saga-test-plan/json/function/checkIfGameFinished",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          Object {
            "type": "game/check-game-finished",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": 1,
          "type": "game/game-finished",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Array [],
          "type": "activeBoardsReducer/set-allowed-boards",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "game/check-game-finished",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "game/check-game-finished",
      },
    },
  ],
}
`;
